### PR TITLE
SALTO-4430 - Salesforce: Add support for arrays and nested fields to unknown_users CV

### PR DESCRIPTION
**Pull request by tomsellek**
The previous PR for this CV was... painfully naive. It assumed we will get changes to instances as changes to the specific type of the inner field of the instance that was actually changed. This not being the case, we need to support validation of nested fields and arrays.

---

---
_Release Notes_: 
N/A

---
_User Notifications_: 
N/A

### CODE REVIEW

File: packages/salesforce-adapter/src/change_validators/unknown_users.ts [hunk 1]
```diff
@@ -59,20 +69,26 @@ type MissingUser = {
...
 
-const userFieldValue = (instance: InstanceElement, fieldName: string): string[] => {
-  if (instance.value[fieldName] === undefined) {
-    return []
+const nestedFieldValue = (container: InstanceElement | Value, fieldName: string): Value => {
+  if (container instanceof InstanceElement) {
+    return resolvePath(container, container.elemID.createNestedID(...fieldName.split(NAMESPACE_SEPARATOR)))
   }
-  return [instance.value[fieldName]]
+  return _.get(container, fieldName)
```
Comment (benisraeladi) [inline packages/salesforce-adapter/src/change_validators/unknown_users.ts:76]: If it's not an `InstanceElement` what else can it be?

**Review by tamtamirr — COMMENTED** [review]
@tomsellek I've started reviewing but I think this PR is too complicated and can be simplified using **walkOnElement**. We could take this F2F if needed :)

File: packages/salesforce-adapter/src/change_validators/unknown_users.ts [hunk 2]
```diff
@@ -37,17 +46,18 @@ type UserFieldGetter = {
...
-  'Approver',
+  'ApprovalProcess',
   'CustomSite',
-  'EmailServicesAddress',
-  'PresenceConfigAssignments',
-  'Users',
+  'EmailServicesFunction',
+  'PresenceUserConfig',
+  'Queue',
 ] as const
```
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/change_validators/unknown_users.ts:61]: Yes. The issue here was my misunderstanding of what the changes look like - I thought we'd get a change on the innermost type that changed in an instance (which doesn't even make sense!) but we get a change on the type of the actual instance.

File: packages/salesforce-adapter/src/change_validators/unknown_users.ts (hunk 1 above)
Comment (tomsellek) [reply packages/salesforce-adapter/src/change_validators/unknown_users.ts:76]: An element in an array (i.e. a `Value`)

File: packages/salesforce-adapter/src/change_validators/unknown_users.ts [hunk 3]
```diff
@@ -59,20 +69,26 @@ type MissingUser = {
   userName: string
 }
 
-const userFieldValue = (instance: InstanceElement, fieldName: string): string[] => {
-  if (instance.value[fieldName] === undefined) {
-    return []
+const nestedFieldValue = (container: InstanceElement | Value, fieldName: string): Value => {
+  if (container instanceof InstanceElement) {
```
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/change_validators/unknown_users.ts]: Ack

**Comment by tomsellek** [comment]
Reworked in https://github.com/salto-io/salto/pull/4563

**Pull request closed without merging**